From 8d01b33d6010a5480a5d7340078df1b5f83e78fb Mon Sep 17 00:00:00 2001
From: Simon McVittie <smcv@debian.org>
Date: Thu, 8 Nov 2018 18:57:03 +0000
Subject: [PATCH] sbuild-createchroot: Don't usrmerge even if it is the
 debootstrap default

Some packages are misbuilt in a merged /usr environment, for example
quilt (#913226) and previously systemd (#843433). For now, let's use
the safe option for buildd chroots.

Closes: #913228
---
 bin/sbuild-createchroot      | 16 ++++++++++++++++
 man/sbuild-createchroot.8.in | 16 ++++++++++++++++
 2 files changed, 32 insertions(+)

diff --git a/bin/sbuild-createchroot b/bin/sbuild-createchroot
index 4d7e74d9..c55bf117 100755
--- a/bin/sbuild-createchroot
+++ b/bin/sbuild-createchroot
@@ -92,6 +92,9 @@ sub setup {
 	'CHROOT_MODE'				=> {
 	    DEFAULT => 'schroot'
 	},
+	'MERGED_USR'				=> {
+	    DEFAULT => 0
+	},
     );
 
     $conf->set_allowed_keys(\%createchroot_keys);
@@ -174,6 +177,15 @@ sub set_options {
 	},
         "command-prefix=s" => sub {
 	    $self->set_conf('COMMAND_PREFIX', $_[1]);
+	},
+	"merged-usr" => sub {
+	    $self->set_conf('MERGED_USR', 1)
+	},
+	"auto-merged-usr" => sub {
+	    $self->set_conf('MERGED_USR', 'auto')
+	},
+	"no-merged-usr" => sub {
+	    $self->set_conf('MERGED_USR', 0)
 	});
 }
 
@@ -309,6 +321,10 @@ push @args, "--keyring=" . $conf->get('KEYRING') if $conf->get('KEYRING');
 push @args, "--no-check-gpg" if defined $conf->get('KEYRING') && $conf->get('KEYRING') eq "";
 push @args, $conf->get('RESOLVE_DEPS') ?
     "--resolve-deps" : "--no-resolve-deps";
+if ($conf->get('MERGED_USR') ne 'auto') {
+	push @args, $conf->get('MERGED_USR') ?
+	    "--merged-usr" : "--no-merged-usr";
+}
 push @args, "$suite", "$target", "$mirror";
 push @args, "$script" if $script;
 
diff --git a/man/sbuild-createchroot.8.in b/man/sbuild-createchroot.8.in
index 94e5661c..5321983b 100644
--- a/man/sbuild-createchroot.8.in
+++ b/man/sbuild-createchroot.8.in
@@ -27,6 +27,7 @@ sbuild\-createchroot \- create sbuild chroot
 .RB [ \-\-chroot-mode=\fIschroot|sudo|unshare\fP ]
 .RB [ \-\-foreign ]
 .RB [ \-\-resolve-deps " \[or] " \-\-no-resolve-deps ]
+.RB [ \-\-merged-usr " \[or] " \-\-no-merged-usr " \[or] " \-\-auto-merged-usr ]
 .RB [ \-\-keep-debootstrap-dir ]
 .RB [ \-\-debootstrap=\fIdebootstrap\fP ]
 .RB [ "\-\-include=\fIpackage1[,package2,[packagen]]\fP" ]
@@ -161,6 +162,21 @@ Download signatures for retrieved \fIRelease\fP files and check them against
 \fIkeyring-file\fP.  By default \fI/etc/apt/trusted.gpg\fP is used.  Set to an
 empty string to disable signature checking.
 .TP
+.BR \-\-merged-usr
+Create a chroot in which \fI/bin\fP, \fI/sbin\fP and \fI/lib*\fP are
+symbolic links to their counterparts in \fI/usr\fP.
+.TP
+.BR \-\-no-merged-usr
+Create a chroot in which \fI/bin\fP, \fI/sbin\fP and \fI/lib*\fP are
+ordinary directories distinct from their counterparts in \fI/usr\fP.
+This is the default.
+.TP
+.BR \-\-auto-merged-usr
+Do not specify whether \fI/bin\fP, \fI/sbin\fP and \fI/lib*\fP are
+symbolic links to their counterparts in \fI/usr\fP. In this case
+debootstrap will use its default behaviour (which is suite-specific).
+This might become the default in a future version of sbuild-createchroot.
+.TP
 .B SUITE
 The distribution to bootstrap (e.g. \[oq]stretch[cq], \[oq]buster\[cq],
 \[oq]bullseye\[cq], \[oq]sid\[cq]).  A complete list may be found in
-- 
2.18.0

